File: exam_at_home/2/boolean_expression.cc
/*
* author: wang-yang
* email: tnst92002@gmail.com
*/
#include <bits/stdc++.h>
using namespace std;

int ord[256] = {0};
vector<char> trans(string s) {
    ord['('] = 7122;
    ord['&'] = 1;
    ord['|'] = 2;
    ord['^'] = 3;
    vector<char> res;
    stack<char> st;

    for (char c:s) {
        if (isalnum(c)) {
            res.emplace_back(c);
        } else if (c=='(') {
            st.push(c);
        } else if (c==')') {
            while (true) {
                c = st.top(); st.pop();
                if (c=='(')
                    break;
                res.emplace_back(c);
            }
        } else {
            while (!st.empty() && ord[st.top()] < ord[c] ) {
                res.emplace_back(st.top());
                st.pop();
            }
            st.push(c);
        }
    }

    while (!st.empty()) {
        res.emplace_back(st.top());
        st.pop();
    }

    return res;
}

bool run(const vector<char> &expr, bool *v) {
    stack<bool> st;

    for (auto c : expr) {
        if (isdigit(c)) {
            st.emplace(c-'0');
        } else if (isalpha(c)) {
            st.emplace(v[c-'A']);
        } else {
            bool R = st.top(); st.pop();
            bool L = st.top(); st.pop();

            switch (c) {
                case '&': L&=R; break;
                case '|': L|=R; break;
                case '^': L^=R; break;
            }

            st.emplace(L);
        }
    }

    return st.top();
}

int main() {
    string buf;
    getline(cin, buf);

    auto expr = trans(buf);

    int T;
    cin >> T;
    while (T--) {
        bool val[5];
        for (int i=0;i<5;++i)
            cin >> val[i];
        cout << !!run(expr, val) << '\n';
    }
}

File: exam_at_home/2/boolean_expression.py
# author: jamie
# email: jinjiedeng.jjd@gmail.com
def Priority (c):
    if   c == '&': return 3
    elif c == '|': return 2
    elif c == '^': return 1
    elif c == '(': return 0

def InfixToPostfix (infix, postfix):
    stack = []
    for c in infix:
        if c == '(':
            stack.append('(')
        elif c == ')':
            while stack[-1] != '(':
                postfix.append(stack.pop())
            stack.pop()
        elif c == '&' or c == '|' or c == '^':
            while len(stack) and Priority(c) <= Priority(stack[-1]):
                postfix.append(stack.pop())
            stack.append(c)
        else:
            postfix.append(c)
    while len(stack):
        postfix.append(stack.pop())

def Evaluate (postfix, value):
    stack = []
    for c in postfix:
        if c == '&' or c == '|' or c == '^':
            rhs = stack.pop()
            lhs = stack.pop()
            if   c == '&': stack.append(lhs & rhs)
            elif c == '|': stack.append(lhs | rhs)
            elif c == '^': stack.append(lhs ^ rhs)
        elif c == '1' or c == '0':
            stack.append(ord(c) - ord('0'))
        else:
            stack.append(value[ord(c) - ord('A')])
    return stack.pop()

if __name__ == "__main__":
    infix = input()
    T = int(input())
    for _ in range(T):
        value = list(map(int, input().split()))
        postfix = []
        InfixToPostfix(infix, postfix)
        print(Evaluate(postfix, value))